x/sequencer/keeper/get_and_set.go

```diff
@@ -0,0 +1,190 @@
+	return k.RollappSequencersByStatus(ctx, rollappId, types.Bonded)
+}
+
+func (k Keeper) AllSequencers(ctx sdk.Context) (list []types.Sequencer) {
+	return k.prefixSequencers(ctx, types.SequencersKeyPrefix)
+}
+
+func (k Keeper) prefixSequencers(ctx sdk.Context, prefixKey []byte) []types.Sequencer {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), prefixKey)
+	it := sdk.KVStorePrefixIterator(store, []byte{})
+
+	defer it.Close() // nolint: errcheck
+
+	ret := []types.Sequencer{}
+	for ; it.Valid(); it.Next() {
+		var val types.Sequencer
+		k.cdc.MustUnmarshal(it.Value(), &val)
+		ret = append(ret, val)
+	}
+
+	return ret
+}
+
+// GetSequencer returns the sentinel sequencer if not found. Use GetRealSequencer if expecting
+// to get a real sequencer.
+func (k Keeper) GetSequencer(ctx sdk.Context, addr string) types.Sequencer {
+	seq, err := k.RealSequencer(ctx, addr)
+	if err != nil {
+		return k.SentinelSequencer(ctx)
+	}
+	return seq
+}
+
+// RealSequencer tries to get a real (non sentinel) sequencer.
+func (k Keeper) RealSequencer(ctx sdk.Context, addr string) (types.Sequencer, error) {
+	store := ctx.KVStore(k.storeKey)
+	b := store.Get(types.SequencerKey(addr))
+	if b == nil {
+		return types.Sequencer{}, types.ErrSequencerNotFound
+	}
+	ret := types.Sequencer{}
+	k.cdc.MustUnmarshal(b, &ret)
+	return ret, nil
+}
+
+func (k Keeper) SequencerByDymintAddr(ctx sdk.Context, addr cryptotypes.Address) (types.Sequencer, error) {
+	accAddr, err := k.dymintProposerAddrToAccAddr.Get(ctx, addr)
+	if err != nil {
+		if errorsmod.IsOf(err, collections.ErrNotFound) {
+			return types.Sequencer{}, gerrc.ErrNotFound
+		}
+		return types.Sequencer{}, err
+	}
+	return k.RealSequencer(ctx, accAddr)
+}
+
+func (k Keeper) AllProposers(ctx sdk.Context) (list []types.Sequencer) {
+	return k.prefixSequencerAddrs(ctx, types.ProposerByRollappKey(""))
+}
+
+func (k Keeper) AllSuccessors(ctx sdk.Context) []types.Sequencer {
+	return k.prefixSequencerAddrs(ctx, types.SuccessorByRollappKey(""))
+}
+
+func (k Keeper) prefixSequencerAddrs(ctx sdk.Context, pref []byte) []types.Sequencer {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), pref)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close() // nolint: errcheck
+	ret := []types.Sequencer{}
+	for ; iterator.Valid(); iterator.Next() {
+		address := string(iterator.Value())
+		seq := k.GetSequencer(ctx, address)
+		ret = append(ret, seq)
+	}
+	return ret
+}
+
+func (k Keeper) GetProposer(ctx sdk.Context, rollapp string) types.Sequencer {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.ProposerByRollappKey(rollapp))
+	if bz == nil {
+		return k.GetSequencer(ctx, types.SentinelSeqAddr)
+	}
+	return k.GetSequencer(ctx, string(bz))
+}
+
+func (k Keeper) GetSuccessor(ctx sdk.Context, rollapp string) types.Sequencer {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.SuccessorByRollappKey(rollapp))
+	if bz == nil {
+		return k.GetSequencer(ctx, types.SentinelSeqAddr)
```

Comment:
how u distinguish between No successer set (no rotation in progress), and sentinal successor?

Reply:
Why do you need to do that?
If you need to do that,  awaitingLastProposerBlock
Can expose in a query if need be